_content/tour/ukr/composition-pollution.article

```diff
@@ -1,86 +1,86 @@
+- Якщо потрібна фінансова допомога, то використовуйте [[https://www.ardanlabs.com/scholarship/][форму для отримання стипендії]]
 
-Interface pollution comes from the fact that people are designing software with
-interfaces instead of discovering them.
+Забруднення інтерфейсами випливає з факту того, що люди створюють дизайн програмного забезпечення за допомогою інтерфейсів
+замість того, щоб відкривати його природно.
 
-** Code Review
+** Перегляд коду
 
-- *Example* *1:* Create Interface Pollution
-- *Example* *2:* Remove Interface Pollution
+- *Example* *1:* Створіть Забруднення Інтерфейсів
+- *Example* *2:* Приберіть Забруднення Інтерфейсів
 
 .play composition/pollution/example1.go
 .play composition/pollution/example2.go
 
-** Interface Pollution
+** Забруднення Інтерфейсами
 
-Interface pollution comes from the fact that people are designing software with
-interfaces instead of discovering them. You should design a concrete solution to the
-problem first. Then you can discover where the program needs to be polymorphic, if
-at all.
+Забруднення Інтерфейсами випливає з факту того, що люди проектують програмне забезпечення за допомогою інтерфейсів
+замість того, щоб відкривати його природно. Спочатку ви повинні створити
+конкретне рішення для проблеми. Потім ви можете зрозуміти де саме програма має буде поліморфною,
+або чи має вона такою бути взагалі.
 
-These are things I’ve heard from other developers.
+Ось які речі я чув від інших розробників.
 
-"I’m using interfaces because we have to use interfaces".
+"Я використовую інтерфейси тому що ми маємо використовувати інтерфейси".
 
-No. We don't have to use interfaces. We use interfaces when it’s practical and
-reasonable to do so. There is a cost of using interfaces: a level of indirection
-and allocation when we store concrete values inside of them. Unless the cost of
-the allocation is worth what I’m gaining by decoupling, you shouldn't be using interfaces.
+Ні. Ми не маємо використовувати інтерфейси. Ми використовуємо інтерфейси тоді, коли в цьому є практичний зміст,
+або коли це доцільно. У використанні інтерфейсів є своя ціна: певний рівень опосередкованості
+та додаткове виділення памʼяті, коли ми зберігаємо значення всередині. Якщо ціна додаткового виділення памʼяті
+не переважає того, що ми отримуємо цим опосередкуванням, то використання інтерфейсів треба уникати.
 
-"I need to be able to test my code so I need to use interfaces".
+"Я маю тестувати свій код, тому мені треба використовувати інтерфейси".
 
-No. you must design my API for the user first, not my test. If the API is not testable,
-You should question if it’s usable. There are different layers of API’s as well. The
-lower level unexported API’s can and should focus on testability. The higher level
-exported API’s need to focus on usability.
+Ні. Ви маєте створювати свій API в першу чергу для користувача, а не для тестів. Якщо API неможливо відтестувати,
```

Comment:
```suggestion
Ні. Ви маєте створювати свій API найперше для користувача, а не для тестів. Якщо API неможливо відтестувати,
```

https://onlinecorrector.com.ua/uk/%D0%BD%D0%B0%D1%81%D0%B0%D0%BC%D0%BF%D0%B5%D1%80%D0%B5%D0%B4-%D0%BF%D0%B5%D1%80%D0%B5%D0%B4%D1%83%D1%81%D1%96%D0%BC-%D1%89%D0%BE%D0%BD%D0%B0%D0%B9%D0%BF%D0%B5%D1%80%D1%88%D0%B5